### DISCUSSION TITLE
[Question][Lecture03] Defining "good state" and "bad state" when checking if a sorting algorithm works correctly

### DISCUSSION THREAD

**Discussion by jaehyun1ee** [prosyslab-classroom/cs424-program-reasoning#385] [Q&A]
I have one question about today's lecture, regarding the definition of a "good state" in the sorting algorithm example.

In lecture03 slide 15, a property "Should return a sorted array (if terminated)" is a safety property.
Then, I wonder how we can define a "good state" and a "bad state" regarding this property.

If we define a "good property" as, "an array is sorted", the algorithm starts with a "bad state", where an input array is not sorted yet.
But we have learned that for safety property, a trace is initially in a good state and we try to prove that it never falls into a "bad state" in finite time.

Then, should we define a "good state" as, "ON RETURN, the array is sorted" in this situation?
And all states that are not return statements should be regarded as "good states"?

Thank you in advance :)

**Comment by KihongHeo** [comment]
Oh, what a coincidence!

See #16 and let me know if you still don't understand.

**Comment by jaehyun1ee** [comment]
Thank you for the explanation!

I think I was confused with the concept of a property and a state definition.
Now, as far as I understood, the definition of a "good" and "bad" state is a tool to prove the safety property, "Should return a sorted array (if terminated)".
i.e., we can devise some state definition that can lead to the proof of the property.
I guess
```
l <= i /\ (forall  j. l <= j < i  -> a[j] != e)
```
would be the case of defining a "good state" to prove the safety property in #16.

Is my understanding correct?

**Comment by KihongHeo** [comment]
Yes. Let me clarify a bit more.

For the safety property in #16, it is enough to show that the "good state" you mentioned is always preserved during the loop.
Here we just focus on the above condition rather than the actual target safety property (i.e., the correctness of the search algorithm), because it is very trivial to prove the target property once the invariant is given.

In general, loop invariant is the most crucial part of verifying safety properties. It is trivial to prove properties about programs without loops. Similarly, you can reduce the verification of the sorting algorithm to find such a "good" state (i.e., a precise enough invariant). 

That is why I just mentioned "good" states for safety properties. More formal proof techniques will be discussed from the next lecture.

**Comment by jaehyun1ee** [comment]
Thank you, things got much clearer.

Looking forward to the next lecture :)
